File: exercises/dotnet/NotificationService/Domain/Identifiers.cs
using SharedObservability;

namespace NotificationService.Domain;

/// <summary>
/// Notification's typed identifier vocabulary. Keys use dotted names aligned
/// with the rest of the workshop services for cross-service correlation.
/// </summary>
public sealed class NotificationContextKey
{
    public static readonly NotificationContextKey OrderId = new("order.id");
    public static readonly NotificationContextKey CustomerId = new("customer.id");
    public static readonly NotificationContextKey CartId = new("cart.id");
    public static readonly NotificationContextKey NotificationId = new("notification.id");

    public string Key { get; }

    private NotificationContextKey(string key) => Key = key;

    /// <summary>
    /// Bind this key to a concrete value, returning an <see cref="IDomainIdentifier"/>.
    /// </summary>
    public IDomainIdentifier Of(string value) => new BoundIdentifier(Key, value);

    /// <summary>
    /// Looks up a <see cref="NotificationContextKey"/> by its dotted key name.
    /// Returns null if the key is not recognized.
    /// </summary>
    public static NotificationContextKey? FromKey(string key) =>
        key switch
        {
            "order.id" => OrderId,
            "customer.id" => CustomerId,
            "cart.id" => CartId,
            "notification.id" => NotificationId,
            _ => null,
        };

    private sealed record BoundIdentifier(string Key, string Value) : IDomainIdentifier;
}

File: exercises/python/notification_service/domain/identifiers.py
"""Domain identifier keys for the Notification bounded context.

Each key maps to a structured observability attribute (span attribute,
structured log field, Kafka header) that propagates across service
boundaries.

NOTIFICATION_ID is purely Notification's. The other keys use the wire
strings agreed with Order via the Kafka header contract --
``domain.order.id``, ``domain.customer.id``, ``domain.cart.id``.
Notification's own enum names them; the wire string IS the cross-context
agreement.
"""

from __future__ import annotations

from enum import Enum

from shared_observability import DomainIdentifier


class _BoundIdentifier:
    """A DomainIdentifier bound to a concrete key and value."""

    __slots__ = ("_key", "_value")

    def __init__(self, key: str, value: str) -> None:
        self._key = key
        self._value = value

    def key(self) -> str:
        return self._key

    def value(self) -> str:
        return self._value


class NotificationContextKey(Enum):
    """Typed identifier vocabulary for the Notification context.

    Shared wire names (order.id, customer.id, cart.id) are intentionally
    aligned with the Order service for cross-service correlation.
    """

    ORDER_ID = "order.id"
    CUSTOMER_ID = "customer.id"
    CART_ID = "cart.id"
    NOTIFICATION_ID = "notification.id"

    def of(self, value: str) -> DomainIdentifier:
        """Bind this key to a concrete value, returning a DomainIdentifier."""
        return _BoundIdentifier(self.value, value)
